### PR TITLE
feat: Implement `await client.start()` to ensure sidecar has started

### PR DESCRIPTION
**Pull request by XavierGeerinck**
# Description

Implement `await client.start()` to ensure sidecar has started

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #268 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x]Extended the documentation

### CODE REVIEW

File: test/e2e/main.http.test.ts [hunk 1]
```diff
@@ -42,6 +42,7 @@ describe('http/main', () => {
     await server.pubsub.subscribe('pubsub-redis', 'test-topic', mockPubSubSubscribe);
 
     // Start server
+    await client.start();
```
Comment (shubham1172) [inline test/e2e/main.http.test.ts:45]: Should we update the docs too?

Comment (shubham1172) [reply test/e2e/main.http.test.ts:45]: And examples.

Comment (XavierGeerinck) [reply test/e2e/main.http.test.ts:45]: Added those, will do a bigger pass later on as well

File: src/implementation/Client/HTTPClient/health.ts [hunk 2]
```diff
@@ -25,16 +25,25 @@ export default class HTTPClientHealth implements IClientHealth {
...
-
-      return true;
+      const result = await this.client.execute(`/metadata`);
+      return !!result;
     } catch (e) {
+      // console.error((e as Error).message);
       return false;
     }
+
+    // try {
```
Comment (shubham1172) [inline src/implementation/Client/HTTPClient/health.ts:35]: Let's either clean this up or add a comment explaining the TODO.

File: src/implementation/Client/HTTPClient/health.ts [hunk 3]
```diff
@@ -25,16 +25,25 @@ export default class HTTPClientHealth implements IClientHealth {
...
-        headers: {
-          'Content-Type': 'application/json',
-        }
-      });
-
-      return true;
+      const result = await this.client.execute(`/metadata`);
+      return !!result;
     } catch (e) {
+      // console.error((e as Error).message);
```
Comment (shubham1172) [inline src/implementation/Client/HTTPClient/health.ts:31]: Do you want to log this or remove it?

File: src/implementation/Client/GRPCClient/GRPCClient.ts [hunk 4]
```diff
@@ -65,4 +65,8 @@ export default class GRPCClient implements IClient {
   async stop(): Promise<void> {
     this.client.close();
   }
+
+  async start(): Promise<void> {
+    this.client.close();
```
Comment (shubham1172) [inline src/implementation/Client/GRPCClient/GRPCClient.ts:70]: Should we `start` here?

Comment (shubham1172) [reply src/implementation/Client/GRPCClient/GRPCClient.ts:70]: Also, should we add this to HTTP Client too?

Comment (XavierGeerinck) [reply src/implementation/Client/GRPCClient/GRPCClient.ts:70]: Goosd catch! It's not called here, so I am removing it

File: src/implementation/Client/GRPCClient/GRPCClient.ts [hunk 5]
```diff
@@ -62,7 +67,69 @@ export default class GRPCClient implements IClient {
...
+    // Dapr will probe every 50ms to see if we are listening on our port: https://github.com/dapr/dapr/blob/a43712c97ead550ca2f733e9f7e7769ecb195d8b/pkg/runtime/runtime.go#L1694
+    // if we are using actors we will change this to 4s to let the placement tables update
+    let isHealthy = false;
+    let isHealthyRetryCount = 0;
+    const isHealthyMaxRetryCount = 60; // 1s startup delay and we try max for 60s
+
+    console.log(`[Dapr-JS][Client] Awaiting Sidecar to be Started`);
+    while (!isHealthy) {
+      console.log(`[Dapr-JS][Client] Waiting till Dapr Sidecar Started (#${isHealthyRetryCount})`);
+      await NodeJSUtils.sleep(Settings.getDaprSidecarPollingDelayMs());
```
Comment (shubham1172) [inline src/implementation/Client/GRPCClient/GRPCClient.ts:116]: Same for this logic. Can we place it in a util and reuse?

File: examples/pubsub/src/index.ts [hunk 6]
```diff
@@ -32,6 +32,7 @@ async function start() {
     process.env.DAPR_HTTP_PORT,
   );
 
+
```
Comment (shubham1172) [inline examples/pubsub/src/index.ts:35]: Extra line?

**Review by shubham1172 — DISMISSED** [review]

File: src/implementation/Client/HTTPClient/HTTPClient.ts [hunk 7]
```diff
@@ -89,16 +94,72 @@ export default class HTTPClient implements IClient {
     this.httpsAgent.destroy();
   }
 
+  async isSidecarStarted(): Promise<boolean> {
+    try {
+      await this.execute(`/metadata`, null, false);
```
[the remark this answers is not in the document]
Comment (XavierGeerinck) [reply src/implementation/Client/HTTPClient/HTTPClient.ts:99]: Yep, I will use this method in the health one

File: src/interfaces/Client/IClient.ts [hunk 8]
```diff
@@ -15,10 +15,12 @@ import CommunicationProtocolEnum from "../../enum/CommunicationProtocol.enum";
 import { DaprClientOptions } from "../../types/DaprClientOptions";
 
 export default interface IClient {
-  getClient(): any; // dependent on implementation
+  getClient(): any; //s dependent on implementation
```
Comment (shubham1172) [inline src/interfaces/Client/IClient.ts:18]: nit typo

**Review by shubham1172 — APPROVED** [review]

**Pull request merged**